### PR TITLE
Corrected and defined behavior of group functions

### LINKED ISSUE #3051: Switch Groups changing to UNDEF
With OH2 GA (might also occur with other versions) I do face an issue with Group:Switch:OR going to "UNDEF" instead of ON or OFF.

I do own some motion sensors (Homematic HM-Sen-MDIR-O-2) and "itemized" them as follows:

```
Group:Switch:OR(ON,OFF) gMotion "Motion detected"
Switch M1 "Motion Sensor #1" <motion> (gMotion) { channel="homematic:..." }
Switch M2 "Motion Sensor #2 " <motion> (gMotion)  { channel="homematic:..." }
```
However, the group's state only is ON or OFF if both sensors have the same state. If their state's differ, the group reports itself as UNDEF.

On motion, the event.log prints

```
[GroupItemStateChangedEvent] - gMotion changed from OFF to UNDEF through M1
[ItemStateChangedEvent     ] - M1 changed from OFF to ON
```
later:

```
[ItemStateChangedEvent     ] - M1 changed from ON to OFF
[GroupItemStateChangedEvent] - gMotion changed from UNDEF to OFF through M1
Source of Issue (issue reported by m0wlheldChristoph Dahlen):
https://community.openhab.org/t/issue-with-group-switch-or-going-to-undef/21403
```

My findings:
If all switches are OFF, the group is OFF
If only one or several are ON, then the group is UNDEF
If all are ON, the group is ON

###################
I checked another group of 30 switches. All have changed from NULL to OFF after starting OH. The group then switches from NULL to OFF too, when the last item of the group switches to OFF.
`gStickyUnreach changed from NULL to OFF through Win_Kind1_StickyUnreach`

So far so good.

But when I change one switch to ON, the group goes UNDEF for some reason.
```
2017-02-18 11:34:46.020 [ItemStateChangedEvent     ] - Heiz_WohnZ_StickyUnreach changed from OFF to ON
2017-02-18 11:34:46.023 [GroupItemStateChangedEvent] - gStickyUnreach changed from OFF to UNDEF through Heiz_WohnZ_StickyUnreach
```

### PR DESCRIPTION
**Pull request by triller-telekom**
A plain group without type does NOT have a function now.
A group with a type uses the EQUALITY function by default
A group with type and function uses the specified function

Fixes #3051

Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 1]
```diff
@@ -36,11 +36,12 @@
     protected GroupFunction function;
 
     public GroupItem(String name) {
-        this(name, null);
+        this(name, null, null);
     }
 
     public GroupItem(String name, GenericItem baseItem) {
-        this(name, baseItem, new GroupFunction.Equality());
+        // this(name, baseItem, new GroupFunction.Equality());
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: ??

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 2]
```diff
@@ -286,7 +287,9 @@ public void stateChanged(Item item, State oldState, State newState) {
     @Override
     public void stateUpdated(Item item, State state) {
         State oldState = this.state;
-        setState(function.calculate(members));
+        if (null != function) {
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: I know, it's a matter of personal taste, but so far we have them the other way round pretty consistently.

**Comment by triller-telekom** [comment]
A group without a function still has a state, but its members do not affect the status of the group. For example: Consider a group with 2 switches swA and swB and I do not have any type or function set. i.e. the items file looks like

```
Group mySwitches
Switch swA (mySwitches)
Switch swB (mySwitches)
```

Then if i trigger `swA` or `swB` the status of `mySwitches` does NOT change, i.e. it stays `NULL`. But if I have a sitemap with:

```
Switch item=mySwitches
```

and I trigger that "group switch", then the status of the group becomes `ON`. And if I trigger that switch again it becomes `OFF`.

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy [hunk 3]
```diff
@@ -117,4 +124,151 @@ class GroupItemTest extends OSGiTest {
...
+        assertTrue change.getItemName().equals(groupItem.getName())
+
+        assertTrue change.getOldItemState().equals(UnDefType.NULL)
+        assertTrue change.getItemState().equals(OnOffType.ON)
+
+        events.clear();
+
+        sw2.setState(OnOffType.ON);
+
+        events.clear();
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy]: as there is no event expected I wouldn't clear `events` here

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy [hunk 4]
```diff
@@ -117,4 +124,151 @@ class GroupItemTest extends OSGiTest {
...
+
+        sw2.setState(UnDefType.UNDEF);
+
+        //wait to see that the event doesn't fire
+        sleep(WAIT_EVENT_TO_BE_HANDLED)
+
+        waitForAssert {
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy]: no need to waitForAssert here

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy [hunk 5]
```diff
@@ -117,4 +124,151 @@ class GroupItemTest extends OSGiTest {
...
+        groupItem.setEventPublisher(publisher)
+        def oldGroupState = groupItem.getState()
+
+        //State changes -> NO change event should be fired
+        member.setState(new RawType())
+
+        //wait to see that the event doesn't fire
+        sleep(WAIT_EVENT_TO_BE_HANDLED)
+
+        waitForAssert {
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy]: no need to waitForAssert here

**Comment by triller-telekom** [comment]
Just did these minor changes on the tests.

**Pull request merged**